src/som/interpreter/StackUtils.java

```diff
@@ -0,0 +1,256 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * <p>
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ * <p>
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package som.interpreter;
+
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.frame.FrameSlotTypeException;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+
+import som.vm.Universe;
+import som.vmobjects.SAbstractObject;
+import som.vmobjects.SBlock;
+import som.vmobjects.SMethod;
+
+
+/**
+ * Arguments.
+ * 0) Interpreter
+ * 2) Invokable
+ * 3) Arguments
+ *
+ * Slots:
+ * 0) Execution stack
+ * 1) Stack pointer
+ * 2) onStack marker
+ */
+public class StackUtils {
+  private enum FrameArguments {
+    INTERPRETER,
+    INVOKABLE,
+    ARGUMENTS
+  }
+
+  public static void initializeStackSlots(final VirtualFrame frame,
+      final FrameSlot executionStackSlot,
+      final SMethod method) {
+
+    int stackLength = (int) method.getMaximumLengthOfStack();
+    SAbstractObject[] stack = new SAbstractObject[stackLength];
+    for (int i = 0; i < stackLength; i++) {
+      stack[i] = Universe.current().nilObject;
+    }
+
+    // TODO - getting the arguments from the frame to put them on stack...
+    // could likely do both when putting them in the stack in the first place
+    // or just don't put them in the frame's arguments, is is useful anyway?
+    final SAbstractObject[] arguments = getCurrentArguments(frame);
+    System.arraycopy(arguments, 0, stack, 0, arguments.length);
+
+    frame.setObject(executionStackSlot, stack);
+    resetStackPointer(frame, method);
+  }
+
+  public static FrameOnStackMarker initializeStackMarkerSlot(VirtualFrame frame,
+      FrameSlot frameOnStackMarkerSlot) {
+    FrameOnStackMarker marker = new FrameOnStackMarker();
+    frame.setObject(frameOnStackMarkerSlot, marker);
+
+    return marker;
+  }
+
+  /***
+   * CURRENT STATE OF THE FRAME.
+   */
+
+  public static SMethod getCurrentMethod(final VirtualFrame frame) {
+    return (SMethod) frame.getArguments()[FrameArguments.INVOKABLE.ordinal()];
+  }
+
+  public static SAbstractObject[] getCurrentArguments(VirtualFrame frame) {
+    return (SAbstractObject[]) frame.getArguments()[FrameArguments.ARGUMENTS.ordinal()];
```

Comment:
Ideally, the arguments are directly in the arguments array and ideally, you can even get rid of the interpreter in the array as well.